notebook/Solution/04 - CNN-Univariate-Multi-Step.ipynb

```diff
@@ -0,0 +1,1474 @@
+   "outputs": [],
+   "source": [
+    "# Create Iterator\n",
+    "train_dataset_shampoo = TensorDataset(trainX_shampoo, trainY_shampoo)\n",
+    "train_iter_shampoo = DataLoader(train_dataset_shampoo,batch_size=batch_size_shampoo,shuffle=False)\n",
+    "\n",
+    "test_dataset_shampoo = TensorDataset(testX_shampoo, testY_shampoo)\n",
+    "test_iter_shampoo = DataLoader(test_dataset_shampoo,batch_size=batch_size_shampoo,shuffle=False)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Input Model\n",
+    "The number of time series feed as the attribute for the model. The other attribute such as kernel_size need to configure in the network configuration itself cause some modification is needed when the number of layer increase."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 144,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "CNN_Multistep(\n",
+      "  (conv1): Conv1d(1, 64, kernel_size=(2,), stride=(1,))\n",
+      "  (poo1): MaxPool1d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)\n",
+      "  (fc1): Linear(in_features=64, out_features=50, bias=True)\n",
+      "  (fc2): Linear(in_features=50, out_features=2, bias=True)\n",
+      ")\n"
+     ]
+    }
+   ],
+   "source": [
+    "model_shampoo = CNN_Multistep(timeseries=trainX_shampoo.shape[1], n_step=n_step_shampoo)\n",
+    "print(model_shampoo.eval())\n",
+    "optimiser_shampoo = torch.optim.SGD(model_shampoo.parameters(), lr=learing_rate_shampoo)\n",
+    "loss_fn_shampoo = nn.MSELoss()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 145,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "CNN_Multistep(\n",
+       "  (conv1): Conv1d(1, 64, kernel_size=(2,), stride=(1,))\n",
+       "  (poo1): MaxPool1d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)\n",
+       "  (fc1): Linear(in_features=64, out_features=50, bias=True)\n",
+       "  (fc2): Linear(in_features=50, out_features=2, bias=True)\n",
+       ")"
+      ]
+     },
+     "execution_count": 145,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "# Xavier weight intialization\n",
+    "def weights_init(m):\n",
+    "    if isinstance(m, nn.Conv1d):\n",
+    "        nn.init.xavier_uniform_(m.weight.data)\n",
+    "        \n",
+    "model_shampoo.apply(weights_init)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 146,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Start Training \n",
+    "train_loss_shampoo,val_loss_shampoo = training(num_epochs_shampoo,train_iter_shampoo,test_iter_shampoo,optimiser_shampoo,loss_fn_shampoo,model_shampoo)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Evaluate\n",
+    "Plot the learning curve to evaluate the loss of each epoch "
```

Comment:
This plot clearly shows overfitting, perhaps it will be helpful to either
1. Include a remark on this, or
2. Show how to revise the model to mitigate overfitting